File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/CS/Direct2D_3D/D2D_D3D_Interop/BitmapUtilities.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using System;
using System.Collections.Generic;
using Microsoft.WindowsAPICodePack.DirectX.WindowsImagingComponent;
using Microsoft.WindowsAPICodePack.DirectX.Direct2D1;
using System.IO;

namespace Microsoft.WindowsAPICodePack.DirectX.Samples
{
    internal class BitmapUtilities
    {
        internal static D2DBitmap LoadBitmapFromFile(
            RenderTarget renderTarget,
            ImagingFactory wicFactory,
            string fileName)
        {

            BitmapDecoder decoder = wicFactory.CreateDecoderFromFileName(fileName, DesiredAccess.Read, DecodeMetadataCacheOption.OnLoad);
            return CreateBitmapFromDecoder(renderTarget, wicFactory, decoder);
        }

        internal static D2DBitmap LoadBitmapFromStream(
            RenderTarget renderTarget,
            ImagingFactory wicFactory,
            Stream ioStream)
        {
            BitmapDecoder decoder = wicFactory.CreateDecoderFromStream(ioStream, DecodeMetadataCacheOption.OnLoad);
            return CreateBitmapFromDecoder(renderTarget, wicFactory, decoder);
        }

        private static D2DBitmap CreateBitmapFromDecoder(RenderTarget renderTarget, ImagingFactory wicFactory, BitmapDecoder decoder)
        {
            BitmapFrameDecode source;
            FormatConverter converter;
            // Create the initial frame.
            source = decoder.GetFrame(0);

            // Convert the image format to 32bppPBGRA -- which Direct2D expects.
            converter = wicFactory.CreateFormatConverter();
            converter.Initialize(
                source.ToBitmapSource(),
                PixelFormats.Pbgra32Bpp,
                BitmapDitherType.None,
                BitmapPaletteType.MedianCut
                );

            // Create a Direct2D bitmap from the WIC bitmap.
            return renderTarget.CreateBitmapFromWicBitmap(
                converter.ToBitmapSource());
        }
    }
}

File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/VB/Direct2D/TextInlineImage/BitmapUtilities.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.


Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports Microsoft.WindowsAPICodePack.DirectX.WindowsImagingComponent
Imports Microsoft.WindowsAPICodePack.DirectX.Direct2D1
Imports System.IO

Namespace Microsoft.WindowsAPICodePack.Samples
	Friend Class BitmapUtilities
		Friend Shared Function LoadBitmapFromFile(ByVal renderTarget As RenderTarget, ByVal wicFactory As ImagingFactory, ByVal fileName As String) As D2DBitmap

            Dim decoder As BitmapDecoder = wicFactory.CreateDecoderFromFileName(fileName, DesiredAccess.Read, DecodeMetadataCacheOption.OnLoad)
			Return CreateBitmapFromDecoder(renderTarget, wicFactory, decoder)
		End Function

		Friend Shared Function LoadBitmapFromStream(ByVal renderTarget As RenderTarget, ByVal wicFactory As ImagingFactory, ByVal ioStream As Stream) As D2DBitmap
            Dim decoder As BitmapDecoder = wicFactory.CreateDecoderFromStream(ioStream, DecodeMetadataCacheOption.OnLoad)
			Return CreateBitmapFromDecoder(renderTarget, wicFactory, decoder)
		End Function

		Private Shared Function CreateBitmapFromDecoder(ByVal renderTarget As RenderTarget, ByVal wicFactory As ImagingFactory, ByVal decoder As BitmapDecoder) As D2DBitmap
			Dim source As BitmapFrameDecode
            Dim converter As FormatConverter
			' Create the initial frame.
			source = decoder.GetFrame(0)

			' Convert the image format to 32bppPBGRA -- which Direct2D expects.
			converter = wicFactory.CreateFormatConverter()
            converter.Initialize(source.ToBitmapSource(), PixelFormats.Pbgra32Bpp, BitmapDitherType.None, BitmapPaletteType.MedianCut)

			' Create a Direct2D bitmap from the WIC bitmap.
			Return renderTarget.CreateBitmapFromWicBitmap(converter.ToBitmapSource())
		End Function
	End Class
End Namespace
